db/sqldb/initiative.go

```diff
@@ -0,0 +1,217 @@
+		if err != nil {
+			return fmt.Errorf("putting initiative: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("updating initiative: %w", err)
+	}
+	return nil
+}
+
+func (d *DB) DeleteInitiative(tx db.Tx, id pacta.InitiativeID) error {
+	err := d.RunOrContinueTransaction(tx, func(db.Tx) error {
+		// TODO(grady) add owner deletions here - where the initiative is the owner of the asset/blob
+		// TODO(grady) do snapshot deletions here
+		err := d.exec(tx, `DELETE FROM initiative_invitation WHERE initiative_id = $1;`, id)
+		if err != nil {
+			return fmt.Errorf("deleting initiative_invitations: %w", err)
+		}
+		err = d.exec(tx, `DELETE FROM initiative_user_relationship WHERE initiative_id = $1;`, id)
+		if err != nil {
+			return fmt.Errorf("deleting initiative_user_relationships: %w", err)
+		}
+		err = d.exec(tx, `DELETE FROM portfolio_initiative_membership WHERE initiative_id = $1;`, id)
+		if err != nil {
+			return fmt.Errorf("deleting portfolio_initiative_memberships: %w", err)
+		}
+		err = d.exec(tx, `DELETE FROM initiative WHERE id = $1;`, id)
+		if err != nil {
+			return fmt.Errorf("deleting initiative: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("performing initiative deletion: %w", err)
+	}
+	return nil
+}
+
+func rowToInitiative(row rowScanner) (*pacta.Initiative, error) {
+	var (
+		pvid pacta.PACTAVersionID
+		lang string
+	)
+
+	i := &pacta.Initiative{}
+	err := row.Scan(
+		&i.ID,
+		&i.Name,
+		&i.Affiliation,
+		&i.PublicDescription,
+		&i.InternalDescription,
+		&i.RequiresInvitationToJoin,
+		&i.IsAcceptingNewMembers,
+		&i.IsAcceptingNewPortfolios,
+		&pvid,
+		&lang,
+		&i.CreatedAt,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("scanning into pacta_version: %w", err)
+	}
+	if pvid != "" {
+		i.PACTAVersion = &pacta.PACTAVersion{ID: pvid}
+	}
+	l, err := pacta.ParseLanguage(lang)
+	if err != nil {
+		return nil, fmt.Errorf("parsing pacta_version language: %w", err)
+	}
+	i.Language = l
+	return i, nil
+}
+
+func rowsToInitiatives(rows pgx.Rows) ([]*pacta.Initiative, error) {
+	return allRows("initiative", rows, rowToInitiative)
+}
+
+func (db *DB) putInitiative(tx db.Tx, i *pacta.Initiative) error {
+	err := db.exec(tx, `
+		UPDATE initiative SET
+			name = $2,
+			affiliation = $3, 
+			public_description = $4,
+			internal_description = $5,
+			requires_invitation_to_join = $6,
+			is_accepting_new_members = $7,
+			is_accepting_new_portfolios = $8,
+			pacta_version_id = $9,
+			language = $10,
+		WHERE id = $1;
+		`, i.ID, i.Name, i.Affiliation, i.PublicDescription, i.InternalDescription, i.RequiresInvitationToJoin, i.IsAcceptingNewMembers, i.IsAcceptingNewPortfolios, i.PACTAVersion.ID, i.Language)
+	if err != nil {
+		return fmt.Errorf("updating initiative writable fields: %w", err)
+	}
+	return nil
+}
+
+// Unlike other mechanisms for creating IDs, initiative IDs are user-specified.
+var initiativeIDRegex = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)
```

Comment:
GC at putting this check at the server layer.

1-liner regexes are still valuable!! Yes, it will take 1ms more to check them (on creation only, a somewhat rare event) but they're infinitely more grokable (IMO) than the code above, and I don't think the time tradeoff is worth it.

Reply:
SG, and lol it's even smaller than that, it's ~1 **micro**second longer to run the regex than the manual check

Reply:
how are computers this fast????